### PR TITLE
Assignment-1

### PR DESCRIPTION
**Pull request by varshilnayak18**
- Screenshots

- Drift Video Link: https://video.drift.com/v/ab4ZvRWtxoVu4xi9mzVvgclmYj3cKTIfdARkwnwpJD20/

### CODE REVIEW

**Comment by varshilnayak18** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 3 |
| code_smells | 14 |
| reliability_rating | D (at least 1 Critical Bug)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

**Comment by varshilnayak18** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 1 |
| code_smells | 3 |
| reliability_rating | C (at least 1 Major Bug)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/assignments/1/src/main/java/org/example/Main.java [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+        logger.logInfo("Top 5 traders:- ");
+        for(Traders t: topTraders){
+            logger.logInfo(t.getFirstName());
+        }
+        List<Traders> bottomTraders = traders.stream().sorted((trader, t1) -> (int) (trader.getPortfolio().getNetProfit() - t1.getPortfolio().getNetProfit())).limit(5).toList();
+        logger.logInfo("Bottom 5 traders:- ");
+        for(Traders t: bottomTraders){
+            logger.logInfo(t.getFirstName());
+        }
+    }
```
Comment (Sagun15) [inline backend/assignments/1/src/main/java/org/example/Main.java:176]: Could have taken user input to print number of traders

File: backend/assignments/1/src/main/java/org/example/Main.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+    static Logging logger = new Logging();
+    static Scanner scanner = new Scanner(System.in);
+    static ExecutorService executorService;
+    static CountDownLatch countDownLatch;
+    static JsonNode jsonNode;
+
+    public static void main(String[] args) {
+
+        //loading csv files
+        Path traderPath = Path.of("/home/lenovo/Desktop/KDU/backend/assignment/javaAssignment1/javaAssignment1/src/main/resources/traders.csv");
```
Comment (Sagun15) [inline backend/assignments/1/src/main/java/org/example/Main.java:28]: Create a separate constants file

File: backend/assignments/1/src/main/java/org/example/ExecuteTransaction.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            countDownLatch.countDown();
+        }
+    }
+
+    private void buyCoin(JsonNode data){
+        String coinName = data.get("coin").asText();
+        long quantity = data.get("quantity").asLong();
+        String walletAddress = data.get("wallet_address").asText();
+        Traders trader = Main.getTraderbyWallet.get(walletAddress);
+        if (trader == null) {
```
Comment (Sagun15) [inline backend/assignments/1/src/main/java/org/example/ExecuteTransaction.java:52]: if(!trader) can suffice.
Also, using Objects.isNull() for null checks is cleaner

File: backend/assignments/1/src/main/java/org/example/Portfolio.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        }
+        netProfit -= quantity*price;
+        coinPrice.put(coinName,(initialPrice+price)*(initialQuantity+quantity)/(initialQuantity+quantity));
+        coinQuantity.put(coinName,initialQuantity+quantity);
+    }
+
+    public void sellCoin(String coinName,long quantity, double currentPrice){
+        long initialQuantity = getCoinVolume(coinName);
+        coinQuantity.put(coinName,initialQuantity-quantity);
+        netProfit += (quantity)*(currentPrice-getCoinPrice(coinName)) - ((quantity)*(currentPrice-getCoinPrice(coinName)));
```
Comment (Sagun15) [inline backend/assignments/1/src/main/java/org/example/Portfolio.java:54]: Won't it return 0

File: backend/assignments/1/src/main/java/org/example/Portfolio.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    public void printPortfolio(){
+        logger.logInfo("org.example.Portfolio details:- ");
+        coinQuantity.forEach((key, value) ->{
+            if(value != 0){
+                logger.logInfo("Coin: " + key + ", Price: " + coinPrice.get(key) + ", Quantity: " + value);
+            }
+        });
+    }
```
Comment (Sagun15) [inline backend/assignments/1/src/main/java/org/example/Portfolio.java:68]: Missing javadocs

**Pull request closed without merging**